core/src/test/java/org/elasticsearch/action/search/CanMatchIT.java

```diff
@@ -0,0 +1,50 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.action.search;
+
+import org.elasticsearch.action.admin.indices.create.CreateIndexRequest;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.index.query.QueryStringQueryBuilder;
+import org.elasticsearch.test.ESIntegTestCase;
+
+public class CanMatchIT extends ESIntegTestCase {
+
+    /*
+     * Previously the can match phase executed on the same thread pool. In cases that a query coordinating node held all the shards for a
+     * query, the can match phase would recurse and end in stack overflow (and this thread would be a networking thread, tying such a thread
+     * up for an extended period of time). This test is for that situation.
+     */
+    public void testAvoidStackOverflow() throws InterruptedException {
```

Comment:
nit - this isn't really a can match test but more an "can we run with lot's of shards on nodes" tests. I think it's good to have but maybe we can fold it into one of the generic search IT suites? maybe `SimpleSearchIT` (although nothing is simple in life ;)